rfcs/text/0011_persistable_state_service.md

```diff
@@ -0,0 +1,340 @@
+    }
+  }
+}
+```
+
+Every plugin that exposes its state should register its `PersistableStateDefinition`. In this definition he needs to expose `migrate` function which can take in any state version and migrate it to the latest state as well as `inject` and `extract` functions, which should be able to inject and extract saved object references from the latest state version.
+
+This same persitableStateDefinition should also be exposed on the plugin contract. On top of that plugin also needs to export type definition for the latest state version. Name of type definition should not include a version number but always be the same. (see the basic example)
+
+Plugins consuming state not owned by them should always use persistableStateDefinion of the registrator plugin. When loading state they need to:
+
+`const state = persistabelStateDefinition.inject(persistableStateDefinition.migrate(loadedState, version), references)`
+
+and when saving state they always need to:
+
+`const { state, references } = persistableStateDefinition.extract(runtimeState)` and store references separatly.
+
+This way consuming plugin can always be sure that its working with the latest state version.
+
+# Server or Client ?
+
+Persistable state service will be available on both, server and client, under same interface. We suggest to register your persistable state definition on both client and server even if you are using just one of those as you cannot know where the consuming side is running.
+
+
+# Consuming and edge cases
+
+When plugin A wants to store some state that it does not own it should check with persistableStateRegistry for a registered persistable state definition and execute `beforeSave` function.
+
+```ts
+  const stateForSaving = persistableStateRegistry.beforeSave(id, myState);
+
+``` 
+
+WARNING: If state id is known at compile time we should rather import the correct utilities directly from the registrator plugin. As using the registry can hide dependencies between plugins when possible plugins should directly depend on plugins providing the state they want to store.
+
+```ts
+  import { extract } from 'src/plugins/pluginOwningState';
+  const stateForSaving = extract(myState);
+```
+
+## Handling corrupt state
+
+In current proposal handling of corrupt state is up to the implementator of PersistableStateDefinition. State incoming to migrate function could not match with version string provided so author should validate the incoming state first.
+
+## Handling enhancements
+
+Related github issue: https://github.com/elastic/kibana/issues/68880
+
+The registrator needs to explicitly support enhancements. Its state should contain an `enhancement` property which is an object where every key represents a state belonging to enhancer.
+
+Drilldown  plugin (enhancer) adds to the state of embeddable (registrator):
+
+```ts
+interface EnhancedDrilldownEmbeddableInput extends EmbeddableInput {
+  enhancements: {
+    drillDowns: { events?: Events[];  }
+  }
+}
```

Comment:
As this is under the registrator's responsibility to handle these 'enhancements' anyway (and not the service/registry), Isn't that a specific implementation from an existing need/structure from our code more than a generic solution that have to be applied everywhere?

Reply:
i agree, this is not something that should be applied everywhere, but its listed here under the examples as Stacey had questions about how that would work, so i just wanted to demonstrate that this is not an issue for persistable state registry